src/expr/src/linear.rs

```diff
@@ -745,57 +745,17 @@ impl MapFilterProject {
         }
         let mut new_expressions = Vec::new();
 
-        // A helper method which memoizes expressions by recursively memoizing their parts.
-        fn memoize_expr(
-            expr: &mut MirScalarExpr,
-            new_scalars: &mut Vec<MirScalarExpr>,
-            projection: &HashMap<usize, usize>,
-            input_arity: usize,
-        ) {
-            match expr {
-                MirScalarExpr::Column(index) => {
-                    // Column references need to be rewritten, but do not need to be memoized.
-                    *index = projection[index];
-                }
-                _ => {
-                    // We should not eagerly memoize `if` branches that might not be taken.
-                    // TODO: Memoize expressions in the intersection of `then` and `els`.
-                    if let MirScalarExpr::If { cond, then, els } = expr {
-                        memoize_expr(cond, new_scalars, projection, input_arity);
-                        // Conditionally evaluated expressions still need to update their
-                        // column references.
-                        then.permute_map(projection);
-                        els.permute_map(projection);
-                    } else {
-                        expr.visit1_mut(|e| memoize_expr(e, new_scalars, projection, input_arity));
-                    }
-                    if let Some(position) = new_scalars.iter().position(|e| e == expr) {
-                        // Any complex expression that already exists as a prior column can
-                        // be replaced by a reference to that column.
-                        *expr = MirScalarExpr::Column(input_arity + position);
-                    } else {
-                        // A complex expression that does not exist should be memoized, and
-                        // replaced by a reference to the column.
-                        new_scalars.push(std::mem::replace(
-                            expr,
-                            MirScalarExpr::Column(input_arity + new_scalars.len()),
-                        ));
-                    }
-                }
-            }
-        }
-
         // We follow the same order as for evaluation, to ensure that all
         // column references exist in time for their evaluation. We could
         // prioritize predicates, but we would need to be careful to chase
         // down column references to expressions and memoize those as well.
         let mut expression = 0;
         for (support, predicate) in self.predicates.iter_mut() {
             while self.input_arity + expression < *support {
+                self.expressions[expression].permute_map(&remaps);
```

Comment:
I'm a little concerned about this, as I am not certain it is equivalent to the above. It may be, but for example the above creates new column references when it memoizes things, and .. I'm not sure what happens. They aren't present in the map, but I would guess that causes a look-up failure rather than a "no worries, I won't rewrite"?

Reply:
On the other hand, maybe it is totally correct? I'm trying to get my head around it more, and the permutation being done *beforehand* seems like it doesn't trip the issues. And the `projection` in the method above being a static reference means as it can't be modified it would only update references to existing columns? Ooof. It's weird that what I need are comments about the changes that shouldn't land in the code if the simplification is indeed correct.